### PR TITLE
KEYCLOAK-16996 User not able to revoke his offline token for directGr…

### PR DESCRIPTION
**Pull request by mposolda**
…ant clients

### CODE REVIEW

**Comment by mposolda** [comment]
@stianst @ssilvert Do you please have a chance to review this?

Before this fix, the "Applications" page contains clients with baseUrl and clients. Clients without baseUrl were filtered. This caused that for clients without baseUrl, it wasn't effectively possible to revoke consent or offlineAccess as users did not see them in the account console.

In this PR, I've just removed the filtering for "baseURL" . IMO since we allow configure clients in the admin console without baseUrl and users are able to login to them, I don't see any reason why the filtering for missing baseUrl should be present in the account console.

**Comment by mposolda** [comment]
@vmuzikar Do you please have a chance to review this? Thanks

**Review by vmuzikar — COMMENTED** [review]
@mposolda If I recall correctly, it [was intentional](https://github.com/keycloak/keycloak/pull/6382#issuecomment-552401731) not to display apps with empty base URL. So maybe we should display those and only if they have direct grant enabled?

**Comment by stianst** [comment]
Looks like there's more fundamental problems here in the implementation. It was supposed to include:

* All clients with consents
* All clients in sessions (both regular and offline)
* Clients not in the above, but where a user may want to see them in the account console for discovery. In this case there should be a base-url, and also if memory serves me right there was supposed to be a flag or something on clients to display them in the account console?

**Comment by stianst** [comment]
The applications page is supposed to list:

1. All clients in sessions (online/offline)
2. All clients with persisted consent
3. Some specific clients that are not in-use to allow users to discover applications

The implementation here is definitively wrong as it filters both 1 and 2, while it should only filter 3. With regards to the exact filtering for 3 we did have a lot of back/forth on how to do this (i.e. apps that have a base-url has it made less sense for folks to discover an app they can't go to, and also we where thinking about requiring at least one client role, but I think we figured that was to expensive). In the end we ended up with a simple specific option for a client to "always include in account console". Not sure I find that the best approach as it doesn't consider what apps an individual user can actually access, but it's a simple approach and it works. With that approach the filtering on base-url becomes a bit mute, as the admin controls directly what applications are included or not.

In summary this fix works for me, but we may want to revisit this at some point in the future to do something better. Especially if we actually at some point implement a way to decide which applications a specific user can or can not login to.

**Review by stianst — APPROVED** [review]

**Review by ssilvert — APPROVED** [review]
The javascript change looks fine.

Will this PR also fix https://issues.redhat.com/browse/KEYCLOAK-18679?

**Comment by stianst** [comment]
@ssilvert yes, it'll fix that one as well so will close it
